File: backend/models/product.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Product 商品基础信息
type Product struct {
	ID          uint           `json:"ID" gorm:"primaryKey"`
	SKU         string         `json:"SKU" gorm:"size:255;uniqueIndex"`
	Name        string         `json:"Name" gorm:"size:255;not null"`
	CategoryID  *uint          `json:"CategoryID" gorm:"default:null"`   // 商品类别ID
	BrandID     *uint          `json:"BrandID" gorm:"default:null"`      // 品牌ID
	Image       string         `json:"Image" gorm:"size:255"`            // 主图
	Images      []string       `json:"Images" gorm:"-"`                  // 多张图片，不存储在数据库中，使用ProductImage关联
	ImagesJSON  string         `json:"-" gorm:"column:images;type:text"` // 兼容旧版本，存储JSON格式的图片列表
	Description string         `json:"Description" gorm:"type:text"`     // 商品描述
	CostPrice   float64        `json:"CostPrice"`                        // 成本价
	RetailPrice float64        `json:"RetailPrice"`                      // 零售价
	Status      bool           `json:"Status" gorm:"default:true"`       // 状态：上架/下架
	CreatedAt   time.Time      `json:"CreatedAt"`
	UpdatedAt   time.Time      `json:"UpdatedAt"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`

	// 关联
	Category      *DictionaryItem  `json:"Category,omitempty" gorm:"foreignKey:CategoryID"` // 商品类别
	Brand         *DictionaryItem  `json:"Brand,omitempty" gorm:"foreignKey:BrandID"`       // 品牌
	ProductImages []ProductImage   `json:"-" gorm:"foreignKey:ProductID"`                   // 商品图片
	Variants      []ProductVariant `json:"Variants,omitempty" gorm:"foreignKey:ProductID"`  // 商品变体
}

// ProductImage 商品图片模型
type ProductImage struct {
	ID        uint           `json:"ID" gorm:"primaryKey"`
	ProductID uint           `json:"ProductID" gorm:"index;not null"`
	URL       string         `json:"URL" gorm:"size:255;not null"`
	Sort      int            `json:"Sort" gorm:"default:0"` // 排序顺序
	CreatedAt time.Time      `json:"CreatedAt"`
	UpdatedAt time.Time      `json:"UpdatedAt"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

// ProductVariant 商品变体（SKU级别）
type ProductVariant struct {
	ID          uint           `json:"ID" gorm:"primaryKey"`
	ProductID   uint           `json:"ProductID" gorm:"not null;index"` // 关联的商品ID
	SKU         string         `json:"SKU" gorm:"size:255;uniqueIndex"` // 变体SKU
	ColorID     *uint          `json:"ColorID" gorm:"default:null"`     // 颜色ID
	SizeID      *uint          `json:"SizeID" gorm:"default:null"`      // 尺码ID
	SeasonID    *uint          `json:"SeasonID" gorm:"default:null"`    // 季节ID
	FabricID    *uint          `json:"FabricID" gorm:"default:null"`    // 面料ID
	Barcode     string         `json:"Barcode" gorm:"size:100"`         // 条形码
	QRCode      string         `json:"QRCode" gorm:"size:255"`          // 二维码数据
	CostPrice   float64        `json:"CostPrice"`                       // 成本价（可能与主商品不同）
	RetailPrice float64        `json:"RetailPrice"`                     // 零售价（可能与主商品不同）
	Status      bool           `json:"Status" gorm:"default:true"`      // 状态：启用/禁用
	CreatedAt   time.Time      `json:"CreatedAt"`
	UpdatedAt   time.Time      `json:"UpdatedAt"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`

	// 关联
	Product *Product        `json:"-" gorm:"foreignKey:ProductID"`               // 商品
	Color   *DictionaryItem `json:"Color,omitempty" gorm:"foreignKey:ColorID"`   // 颜色
	Size    *DictionaryItem `json:"Size,omitempty" gorm:"foreignKey:SizeID"`     // 尺码
	Season  *DictionaryItem `json:"Season,omitempty" gorm:"foreignKey:SeasonID"` // 季节
	Fabric  *DictionaryItem `json:"Fabric,omitempty" gorm:"foreignKey:FabricID"` // 面料
}

File: mobile/lib/models/product.dart
class Product {
  final int id;
  final String sku;
  final String name;
  final int? categoryId;
  final int? brandId;
  final String image;
  final List<String> images;
  final String description;
  final double costPrice;
  final double retailPrice;
  final bool status;
  final String createdAt;
  final String updatedAt;
  final Map<String, dynamic>? category;
  final Map<String, dynamic>? brand;
  final List<ProductVariant>? variants;
  
  Product({
    required this.id,
    required this.sku,
    required this.name,
    this.categoryId,
    this.brandId,
    required this.image,
    required this.images,
    required this.description,
    required this.costPrice,
    required this.retailPrice,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.category,
    this.brand,
    this.variants,
  });
  
  factory Product.fromJson(Map<String, dynamic> json) {
    List<String> imagesList = [];
    if (json['Images'] != null) {
      imagesList = List<String>.from(json['Images']);
    }
    
    List<ProductVariant>? variantsList;
    if (json['Variants'] != null) {
      variantsList = List<ProductVariant>.from(
        json['Variants'].map((v) => ProductVariant.fromJson(v))
      );
    }
    
    return Product(
      id: json['ID'],
      sku: json['SKU'],
      name: json['Name'],
      categoryId: json['CategoryID'],
      brandId: json['BrandID'],
      image: json['Image'] ?? '',
      images: imagesList,
      description: json['Description'] ?? '',
      costPrice: json['CostPrice']?.toDouble() ?? 0.0,
      retailPrice: json['RetailPrice']?.toDouble() ?? 0.0,
      status: json['Status'],
      createdAt: json['CreatedAt'],
      updatedAt: json['UpdatedAt'],
      category: json['Category'],
      brand: json['Brand'],
      variants: variantsList,
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'ID': id,
      'SKU': sku,
      'Name': name,
      'CategoryID': categoryId,
      'BrandID': brandId,
      'Image': image,
      'Images': images,
      'Description': description,
      'CostPrice': costPrice,
      'RetailPrice': retailPrice,
      'Status': status,
      'CreatedAt': createdAt,
      'UpdatedAt': updatedAt,
      'Category': category,
      'Brand': brand,
      'Variants': variants?.map((v) => v.toJson()).toList(),
    };
  }
}

class ProductVariant {
  final int id;
  final int productId;
  final String sku;
  final int? colorId;
  final int? sizeId;
  final int? seasonId;
  final int? fabricId;
  final String barcode;
  final String qrCode;
  final double costPrice;
  final double retailPrice;
  final bool status;
  final Map<String, dynamic>? color;
  final Map<String, dynamic>? size;
  final Map<String, dynamic>? season;
  final Map<String, dynamic>? fabric;
  
  ProductVariant({
    required this.id,
    required this.productId,
    required this.sku,
    this.colorId,
    this.sizeId,
    this.seasonId,
    this.fabricId,
    required this.barcode,
    required this.qrCode,
    required this.costPrice,
    required this.retailPrice,
    required this.status,
    this.color,
    this.size,
    this.season,
    this.fabric,
  });
  
  factory ProductVariant.fromJson(Map<String, dynamic> json) {
    return ProductVariant(
      id: json['ID'],
      productId: json['ProductID'],
      sku: json['SKU'],
      colorId: json['ColorID'],
      sizeId: json['SizeID'],
      seasonId: json['SeasonID'],
      fabricId: json['FabricID'],
      barcode: json['Barcode'] ?? '',
      qrCode: json['QRCode'] ?? '',
      costPrice: json['CostPrice']?.toDouble() ?? 0.0,
      retailPrice: json['RetailPrice']?.toDouble() ?? 0.0,
      status: json['Status'],
      color: json['Color'],
      size: json['Size'],
      season: json['Season'],
      fabric: json['Fabric'],
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'ID': id,
      'ProductID': productId,
      'SKU': sku,
      'ColorID': colorId,
      'SizeID': sizeId,
      'SeasonID': seasonId,
      'FabricID': fabricId,
      'Barcode': barcode,
      'QRCode': qrCode,
      'CostPrice': costPrice,
      'RetailPrice': retailPrice,
      'Status': status,
      'Color': color,
      'Size': size,
      'Season': season,
      'Fabric': fabric,
    };
  }
}
